Include/cpython/pystate.h

```diff
@@ -149,6 +149,13 @@ struct _ts {
 
     struct _py_trashcan trash;
 
+    /* Tagged pointer to top-most critical section, or zero if there is no
+     * active critical section. Critical sections are only used in
+     * `--disable-gil` builds (i.e., when Py_NOGIL is defined to 1). In the
+     * default build, this field is always zero.
+     */
+    uintptr_t critical_section;
```

Comment:
It isn't quite clear enough to me so far what the relationship is between a critical section and a thread/tstate.  Does it have to do with reentrancy or is it strictly for the implicit behavior relative to attach/detach?

FWIW, in my mind I'm imagining a critical section as applying globally, i.e. at the level of `PyInterpreterState`.  Since it's tied to per-object mutexes, that's exactly what's going on here.  My naive mental model hasn't factored in anything thread-related, other than that locks are involved.

Reply:
Critical sections are basically a deadlock avoidance layer on top of per-object locks. Deadlocks (and deadlock avoidance) conceptually happen at the thread level.